gap/PolygonalComplexes/constructors.gi

```diff
@@ -459,13 +459,121 @@ __SIMPLICIAL_IntSetConstructor("VerticesInFaces", __SIMPLICIAL_AllTypes,
+            vertices, vertexPaths, findPredecessorBfs, vertexPredsMap, vertexRotationCycleMap,
+            orientedDescr, pair, vertex, predVertex, predRotationCycle, rotationCycle,
+        # BFS related variables:
+            queue, visited, referenceVertex;
+
+        numFaces := Length(Faces(surface));
+
+        if Length(commonDescr) = 1 or numFaces = 1 then
+            return commonDescr;
+        fi;
+
+        # Find the first bound index in commonDescr
+        for i in [1..Length(commonDescr)] do
+            if not IsBound(commonDescr[i]) then
+                continue;
+            fi;
+
+            firstBoundIndex := i;
+            break;
+        od;
+
+        # For each vertex find all vertex adjancies as [vertex, face] pairs
+        vertexAdjencies := List(commonDescr, x -> []);
+        for face in [1..numFaces] do
+            vertices := VerticesOfFace(surface, face);
+            for vertex in vertices do
+                for i in VerticesOfFace(surface, face) do
+                    if not i in vertexAdjencies[vertex] then
+                        Add(vertexAdjencies[vertex], [i, face]);
+                    fi;
+                od;
+            od;
+        od;
+
+        # Perform BFS to find the predecessor of each vertex
+        queue := [firstBoundIndex];
+        visited := List(commonDescr, x -> false);
+        visited[firstBoundIndex] := true;
+        vertexPredsMap := List(commonDescr, x -> []);
+
+        while Length(queue) > 0 do
+            # Dequeue front
+            referenceVertex := Remove(queue, 1);
+
+            # Process each vertex adjancy pair of referenceVertex
+            for pair in vertexAdjencies[referenceVertex] do
+                vertex := pair[1];
+                face := pair[2];
+
+                if not visited[vertex] then
+                    visited[vertex] := true;
+
+                    # Add predecessor pair
+                    vertexPredsMap[vertex] := [referenceVertex, face];
+
+                    # Enqueue neighbour vertex
+                    Add(queue, vertex);
+                fi;
+            od;
+        od;
+
+        # Provide a reference rotation for vertex firstBoundIndex
+        vertexRotationCycleMap := List(commonDescr, x -> ());
+        if IsList(commonDescr[firstBoundIndex]) then
+            vertexRotationCycleMap[firstBoundIndex] := CycleFromList(commonDescr[firstBoundIndex]);
+        else
+            vertexRotationCycleMap[firstBoundIndex] := commonDescr[firstBoundIndex];
+        fi;
+
+        orientedDescr := commonDescr;
+
+        # Iterate over all paths and match to predecessor vertex rotation
+        for vertex in [2..Length(vertexPredsMap)] do
+            if not IsBound(vertexPredsMap[vertex]) then
+                continue;
+            fi;
+
+            face := vertexPredsMap[vertex][2];
+            predVertex := vertexPredsMap[vertex][1];
+            predRotationCycle := vertexRotationCycleMap[predVertex];
```

Comment:
Is it always the case that predVertex<vertex? Since otherwise vertexRotationCycleMap for predVertex wouldn't be defined or?
Is it possible to compute the correct orientation of the cycles directly above in the while loop? Maybe this would be more efficient and you can assume that you know the orientation of the predecessor vertex already.